content/post/078-go-ipfs-0.4.23.md

```diff
@@ -0,0 +1,84 @@
+
+### 🔁 Fixed Resursive Resolving of dnsaddr Multiaddrs
+
+Imagine this: rolling out a brand spanking new set of bootstrap nodes only to discover their new addresses are not resolvable. I know right - bad news bears 🐻!?
+
+Ok, "bear" this in mind: you want to keep all your DNS TXT records [below 512 bytes to avoid UDP fragmentation](https://serverfault.com/questions/840241/do-dns-queries-always-travel-over-udp), otherwise you'll get a truncated reply and have to connect with TCP to get all the records. If you have lots of dnsaddr TXT records then it can be more efficient to use recursive resolving than to get a truncated reply and go through the famous 18-way SYN, SYN-ACK ACK, ACK-SYN, ACK-ACK (...etc, etc) TCP handshake, not to mention the fact that go-ipfs will not even try to fallback to TCP 😅.
+
+Anyway, long story short. We fixed that. You're welcome.
+
+### 📻 Retuned Connection Manager
+
+The Connection Manager has been tuned to better prioritise existing connections by not counting new connections in the "grace" period towards connection limits. New connections are like new friends. You can't hang out with everyone all the time, I mean, it just gets difficult to book a resturant after a while.
+
+You also wouldn't stop being friends with Jane just because you met Sarah _once_ on the train. You and Jane have history, think of everything you've been through. Remember that time when Jane's dog, Dave, ran away? I know, it's a weird name for a dog, I mean who gives a human name to a dog anyway, but I guess that's one of the reasons you like Jane. Anyway, she lost her dog and you both looked all around town for it, you were about to give up but then you heared faint wimpering as you were walking back to the house. Dave had somehow managed to fall into the old abandoned well!
+
+You see?! History! ...and, erh, what was I saying? Oh yeah, Connection Manager - new connections don't cause us to close useful, existing connections (like Jane), but it does mean you'll keep more connections in total. Maybe consider reducing the `HighWater` setting in your config.
+
+### 🍖 Reduced Relay Related DHT Spam
+
+When `AutoRelay` was enabled, your IPFS node basically spent all it's time searching the DHT for relays, like a fly to a honey trap. I say that because every other IPFS node was doing the same and if you had `AutoRelay` and `RelayHop` enabled then, erm, DoS happens. Bad times.
+
+So, instead of spamming the DHT the whole time trying to find random, potentially poor relays, IPFS is now using a pre-defined set of autorelays. I mean, try to tell me that _doesn't_ make sense.
+
+### 🐾 Better Bitswap
+
+Joe has the rare shiny collectable card but your shoes have holes in them and he's standing in the rain on the other side of the playground 💦🕺🏻. This is exactly what bitswap is like when you have a bug that stops you from connecting to providers. Ok, I mean, there's a few differences, but it's basically the same.
```

Comment:
can we clarify what the bug was re providers? was it all providers or just wet ones? ;P A little more on the counterfactual here can shed a little more light on what actually will be changed

Reply:
The bug: we'd cancel the context we were using to connect to providers once we found enough providers but before we finished connecting to them. In many cases, provider record queries will continue until a timeout. However, when we find a _lot_ of providers, we'll finish the query early and may not finish connecting to peers.

I.e., this was a pretty terrible bug.

Reply:
Not sure I understand why it's bad to cancel the context before we connect to all the providers?

Reply:
In less go terms: We'll abort the connection attempts.

Reply:
That is, we find providers and connect to them in parallel. However, once we've found enough providers, we'd _cancel_ any ongoing attempt to connect these providers and walk away.

In other words:

> Joe has the rare shiny collectable card you've been hunting for forever (since yesterday). He's right over there on the other side of the playground. But now that you've found what you're looking for, you're so excited you forget what you were doing and start looking again.
>
> This is exactly what bitswap is like when you have a bug where you stop trying to connect to providers once you've found enough of them. Specifically, if we found _enough_ providers (100) or timed out the provider request, bitswap would cancel any in-progress connection attempts to providers and walk away.

Reply:
lolz - love it.